### DISCUSSION TITLE
Multiple instances of a specific module

### DISCUSSION THREAD

**Discussion by eddan8** [PrismLibrary/Prism#2367] [WPF] [answered]
**Description**
I need to create multiple instances of a specific module. Theses modules should in turn have their own viewmodels displaying different data. 

An example of this would be an instance [A] of a "Link Module", that would contain a certain collection of website links, and another instance [B] that is also a "Link Module", that would contain a different set of links. 

**Issue**
I can create multiple instances of the module project and inject the proper values (diffferent links) into each viewmodel. My problem is that when I navigate to a specific region and view, the viewmodel values in the view are not updated correctly.

**Github repo for demo project**
https://github.com/eddan8/PrimsModuleInstance

The problem can be recreated by clicking the "Generic View" card to open the main window and then try to navigate between "Internal Links" and "External Links". The view is not updated with the correct data.

**Basic Information:**
Target framework: .NET Framework 4.7.2
Prism verion: 8.0.0.1909
IoC Container: DryIoc

**Information to quickly navigate the solution**

**Projects:**
- ModuleInstancesTest.ApplicationShell
- ModuleInstancesTest.Module.LinkCollectionModule

**Project setup:**

ModuleInstancesTest.ApplicationShell

_App.xaml.cs_

I am overriding InitializeModules() to iterate and load each module instance. The modules are specified in a data service with dummy data, i.e. links. In the InitializeModules() method I also use the "_containerRegistry.RegisterInstance" to register a specific viewmodel that I later inject in my MainLinkCollectionViewModel located in the LinkCollectionModule project. Debugging the the project shows that the injection works fine. So far so good. 

ModuleInstancesTest.Module.LinkCollectionModule

_LinkCollectionModule.cs_

Here I first add my views to my "navigation" regions. Then I register the MainLinkCollectionView with the MainWindowRegion, which is the region where the links should be presented.

```
public void OnInitialized(IContainerProvider containerProvider)
{
    _regionManager.AddToRegion("CardsNavigationRegion", containerProvider.Resolve<CardLinkCollectionView>());
    _regionManager.AddToRegion("MainWindowNavigationRegion", containerProvider.Resolve<NavItemLinkCollectionView>());
    _regionManager.RegisterViewWithRegion("MainWindowRegion", typeof(MainLinkCollectionView));
}
```

In this file I also register my view (MainLinkCollectionView) with a unique name (the name is a GUID formatted as a string and is the same name as the module) for navigation. I also register the associated viewmodel (MainLinkCollectionViewModel) and inject the correct data.

```
public void RegisterTypes(IContainerRegistry containerRegistry)
        {
            containerRegistry.RegisterForNavigation<MainLinkCollectionView>(ModuleName);

            ViewModelLocationProvider.Register<MainLinkCollectionView>(() =>
            {
                return new MainLinkCollectionViewModel(Mapping);
            });

        }
```

NavItemLinkCollectionViewModel

This class contains a method for navigating to the specific view. Here I can see that the ModuleName variable is correct and should correspond to the: 

```
containerRegistry.RegisterForNavigation<MainLinkCollectionView>(ModuleName);"
```

...in the _LinkCollectionModule_ , but the view is not updated with the correct viewmodel.

```
private void performOpenMainView()
{
    _regionManager.RequestNavigate("MainWindowRegion", ModuleName);
}
```

Any ideas of what I might do wrong here? :)

/Peter

**Comment by brianlagunas** [comment, marked as answer]
This app is extremely overcomplicated. There is a lot of improvements that need to be made and a lot of code that needs to be simplified.  This would require a complete architecture review and restructure. Unfortunately, this falls outside the level of community support I provide and would require the use of my paid services.

**Reply by eddan8** [reply]
Thanks for your reply Brian! :) (and for your great Pluralsight courses as well!) 

I see your point. I will go back to the drawingboard and rethink the architechture, problably overcomplicating the whole thing just as you mention. 

/Peter
